### PR TITLE
Migrate to AWS SDK v2

### LINKED ISSUE #430: Migration to AWS SDK V2
AWS released the V2 of their SDK. I just learned about it recently.

https://aws.amazon.com/blogs/developer/aws-sdk-for-go-version-2-general-availability/

They claim that it `offers significant performance improvements in CPU and memory utilization over version 1`.

Quite importantly, from a quick glance and a look at some problems we had, it looks like developer experience is much better than V1:
- sensible defaults for many things
- the API looks simpler and easier to use

Is this worth looking into, or is it something we don't want to do?

### PR DESCRIPTION
**Pull request by xenosdio**
## Which problem is this PR solving?

Resolves #430.

## Short description of the changes

Update Golang AWS SDK to v2. This introduces breaking changes on the `SQS` component and v2 of the `SNS` and `SQS` clients.

Examples have also been updated.

Useful material/documentation, regarding migrating from AWS v1 to v2:
* https://aws.github.io/aws-sdk-go-v2/docs/migrating/
* https://aws.github.io/aws-sdk-go-v2/docs/configuring-sdk/endpoints/
* https://aws.github.io/aws-sdk-go-v2/docs/unit-testing/
* https://aws.github.io/aws-sdk-go-v2/docs/making-requests/

### CODE REVIEW

**Review by pkritiotis — DISMISSED** [review]
Congrats on your first contribution 🥳 🚀 
LGTM

**Comment by mantzas** [comment]
@xenosdio congrats on the first one. Since we are in breaking change mood, should we not remove the old implementations and keep only one e.g. client v2 should be the only and obvioulsy removed from v2. WDYT? /cc @pkritiotis

**Comment by xenosdio** [comment]
> @xenosdio congrats on the first one. Since we are in breaking change mood, should we not remove the old implementations and keep only one e.g. client v2 should be the only and obvioulsy removed from v2. WDYT? /cc @pkritiotis

@mantzas, yup will do. Does this also mean removing the async SQS component? It is currently deprecated, too.

**Comment by mantzas** [comment]
> > @xenosdio congrats on the first one. Since we are in breaking change mood, should we not remove the old implementations and keep only one e.g. client v2 should be the only and obvioulsy removed from v2. WDYT? /cc @pkritiotis
> 
> @mantzas, yup will do. Does this also mean removing the async SQS component? It is currently deprecated, too.

I believe this is a good opportunity to break it now and avoid having to break it again.

**Review by pkritiotis — DISMISSED** [review]
Nice cleanup 🚀 
lgtm

File: client/sqs/publisher.go [hunk 1]
```diff
@@ -6,69 +6,70 @@ import (
...
+		},
+		[]string{"queue", "success"},
+	)
+	prometheus.MustRegister(publishDurationMetrics)
 }
 
-// TracedPublisher is an implementation of the Publisher interface with added distributed tracing capabilities.
-type TracedPublisher struct {
-	api sqsiface.SQSAPI
+type API interface {
```
Comment (mantzas) [inline client/sqs/publisher.go:43]: Somebody might argue here that it is not the concern of the publisher to create queues and ReceiveMesages.
The publisher should only care about publishing.
Propose to remove CreateQueue and ReceiveMessage and keep a single method interface.

File: component/sqs/component.go [hunk 2]
```diff
@@ -90,29 +91,38 @@ type retry struct {
...
+	pollWaitSeconds   int32
+	visibilityTimeout int32
 }
 
 type stats struct {
 	interval time.Duration
 }
 
+type API interface {
+	CreateQueue(ctx context.Context, params *sqs.CreateQueueInput, optFns ...func(*sqs.Options)) (*sqs.CreateQueueOutput, error)
```
Comment (mantzas) [inline component/sqs/component.go:104]: Same here is the component interested in all these methods? We should focus only on the methods that the component needs.

**Review by mantzas — CHANGES_REQUESTED** [review]
Really nice. Added 2 things for change.
We need also to update the Acknowledgements makrdown in the docs folder to point the right package.

Another thing that we need for the release is a procedure on how to upgrade.

**Comment by xenosdio** [comment]
> Another thing that we need for the release is a procedure on how to upgrade.

@mantzas do you mean documentation on how to upgrade from `aws-sdk-go` v1 to v2 for the patron component/client consumers?

File: client/sqs/publisher.go (hunk 1 above)
Comment (xenosdio) [reply client/sqs/publisher.go:43]: Updated, PTAL. The API interface was used in other testing modules. An interface is present solely for each instance now.

File: component/sqs/component.go (hunk 2 above)
Comment (xenosdio) [reply component/sqs/component.go:104]: Updated, PTAL.

**Comment by mantzas** [comment]
> > Another thing that we need for the release is a procedure on how to upgrade.
> 
> @mantzas do you mean documentation on how to upgrade from `aws-sdk-go` v1 to v2 for the patron component/client consumers?

Yes, this was feedback that we had in the past for breaking changes if they are trivial. If they are we should be fine.
What we do in general is telling people to look at the examples and hint then about change.
Usually we make this part of the release description.
Since you have migrated the existing examples, do you feel that the migration is trivial enough to give just some hints?

**Comment by xenosdio** [comment]
> > > Another thing that we need for the release is a procedure on how to upgrade.
> > 
> > 
> > @mantzas do you mean documentation on how to upgrade from `aws-sdk-go` v1 to v2 for the patron component/client consumers?
> 
> Yes, this was feedback that we had in the past for breaking changes if they are trivial. If they are we should be fine. What we do in general is telling people to look at the examples and hint then about change. Usually we make this part of the release description. Since you have migrated the existing examples, do you feel that the migration is trivial enough to give just some hints?

I have added a guideline draft in the SQS component's readme file. Could you please take a look? Is this something we would want?

**Review by mantzas — DISMISSED** [review]

**Comment by xenosdio** [comment]
Fyi, as discussed privately, we will be adding a `BREAKING.md` file mentioning all the related breaking changes from version to version.

Created a first draft with the AWS SDK migration. PTAL.

**Review by mantzas — APPROVED** [review]

File: BREAKING.md [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+  return nil, err
+ }
+
+ api := sqs.NewFromConfig(cfg)
+
+ return api, nil
+}
+```
+
+> A more detailed documentation on migrating can be found [here](https://aws.github.io/aws-sdk-go-v2/docs/migrating).
```
Comment (pkritiotis) [inline BREAKING.md:75]: Very nice 🚀

**Review by pkritiotis — APPROVED** [review]

**Pull request merged**